src/metabase/models/permissions.clj

```diff
@@ -221,93 +221,169 @@
   1. Any character other than a slash
   2. A forward slash, escaped by a backslash: `\\/`
   3. A backslash escaped by a backslash: `\\\\`"
-  (u.regex/rx (or #"[^\\/]" #"\\/" #"\\\\")))
-
-(def ^:private path-regex
+  (u.regex/rx [:or #"[^\\/]" #"\\/" #"\\\\"]))
+
+(def ^:private data-rx->data-kind
+  {      #"db/\d+/"                                                                  :dk/db
+   [:and #"db/\d+/" "native" "/"]                                                    :dk/db-native
+   [:and #"db/\d+/" "schema" "/"]                                                    :dk/db-schema
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/"]                                  :dk/db-schema-name
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/"]                       :dk/db-schema-name-and-table
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/" "read/"]               :dk/db-schema-name-table-and-read
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/" "query/"]              :dk/db-schema-name-table-and-query
+   [:and #"db/\d+/" "schema" "/" path-char "*" "/table/\\d+/" "query/" "segmented/"] :dk/db-schema-name-table-and-segmented})
```

Comment:
this could be more performant, yes. But IMO the clarity here is better, at least until we are finished upgrading perm-path formats.